### PR TITLE
build(docs-infra): support hiding constructors of injectable classes

### PR DESCRIPTION
**Pull request by petebacondarwin**
Class that are injectable often have constructors that should not be
called by the application developer. It is the responsibility of the
injector to instantiate the class and the constructor often contains
private implementation details that may need to change.

### CODE REVIEW

**Review by gkalpak — APPROVED** [review]
You might want to also update [overview-dump.template.html][1].

[1]: https://github.com/petebacondarwin/angular/blob/aca3297da09eff83e55605fa6381f9d5e86bd14f/aio/tools/transforms/templates/overview-dump.template.html#L97

**Comment by gkalpak** [comment]
On second thought, I find the tag name a little confusing. `docsNotRequired` might also mean "you don't need to add docs" (for whatever reason 😛). It is not clear that it will be hidden in docs.

Maybe something more assertive, like `hideInDocs` or `excludeFromDocs`?

**Comment by petebacondarwin** [comment]
How about even more descriptive and specific: `@injectableConstructor`?
Or... we could automatically hide the constructor if the following is true:

1) the class has `@Injectable` decorator
2) the constructor has no description

What do you think?

**Comment by gkalpak** [comment]
If it is true that all `@Injectable` constructors should not appear in docs, then option (2) sounds perfect :smiley:

**Comment by petebacondarwin** [comment]
I rewrote this PR so you don't need to annotate the class constructor to make it hidden. I also noticed that `Component` and other decorators also indicate that the class is injectable so I am removing their decorators too. See the commit message for details of the approach.

**Comment by petebacondarwin** [comment]
@gkalpak PTAL due to the complete rewrite

File: aio/tools/transforms/angular-api-package/processors/removeInjectableConstructors.spec.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  it('should be available on the injector', () => {
+    const dgeni = new Dgeni([testPackage('angular-api-package')]);
+    const injector = dgeni.configureInjector();
+    const processor = injector.get('removeInjectableConstructors');
+    expect(processor.$process).toBeDefined();
+    expect(processor.$runAfter).toEqual(['processing-docs']);
+    expect(processor.$runBefore).toEqual(['docs-processed']);
+  });
+
+  it('should remove undocumented constructors from docs that have an "Injectable" decorator on it', () => {
```
Comment (gkalpak) [inline aio/tools/transforms/angular-api-package/processors/removeInjectableConstructors.spec.js]: _it_ --> _them_ 😇

File: aio/tools/transforms/angular-api-package/processors/removeInjectableConstructors.spec.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+const Dgeni = require('dgeni');
+
+describe('removeInjectableConstructors processor', () => {
+
+  it('should be available on the injector', () => {
+    const dgeni = new Dgeni([testPackage('angular-api-package')]);
+    const injector = dgeni.configureInjector();
+    const processor = injector.get('removeInjectableConstructors');
+    expect(processor.$process).toBeDefined();
+    expect(processor.$runAfter).toEqual(['processing-docs']);
```
Comment (gkalpak) [inline aio/tools/transforms/angular-api-package/processors/removeInjectableConstructors.spec.js]: Shouldn't this be `['processing-docs', 'splitDescription']`?

File: aio/tools/transforms/angular-api-package/processors/removeInjectableConstructors.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+module.exports = function removeInjectableConstructors() {
+  return {
+    $runAfter: ['processing-docs', 'splitDescription'],
+    $runBefore: ['docs-processed'],
+    injectableDecorators: ['Injectable', 'Directive', 'Component', 'Pipe'],
```
Comment (gkalpak) [inline aio/tools/transforms/angular-api-package/processors/removeInjectableConstructors.js]: What about `NgModule`?

File: aio/tools/transforms/angular-api-package/processors/removeInjectableConstructors.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    $process(docs) {
+      docs.forEach(doc => {
+        if (
+          doc.constructorDoc &&
+          !doc.constructorDoc.shortDescription &&
```
Comment (gkalpak) [inline aio/tools/transforms/angular-api-package/processors/removeInjectableConstructors.js]: Can a `constructorDoc` have a `description` but no `shortDescription`?

**Review by gkalpak — CHANGES_REQUESTED** [review]

File: aio/tools/transforms/angular-api-package/processors/removeInjectableConstructors.js (hunk 4 above)
Comment (petebacondarwin) [reply aio/tools/transforms/angular-api-package/processors/removeInjectableConstructors.js]: Not if it runs after the `splitDescription` processor.

**Comment by jbogarthyde** [comment]
This solution makes the doc writer responsible for NOT including a doc comment (or including an empty one) on the constructor for a service. Is that what you want? If so, I will add it to the guidelines.

An explicit tag would have the advantage of showing that the choice not to document is explicit, and not a side-effect of incomplete documentation.

**Comment by petebacondarwin** [comment]
Actually they could still add an explicit tag, e.g. `@docsNotRequired`, which would have no effect but at least it would be a marker to the reviewer? This tag could even have a boilerplate comment. For example:

```
@docsNotRequired this class is `Injectable` so the constructor is a private implementation.
```

As it stands we would not fail if this were missing but perhaps it could become enforceable later.

**Comment by jbogarthyde** [comment]
I like that with the boilerplate -- it would make people think about what they are doing, and provide an explanation for users who are looking at the code.

**Review by gkalpak — APPROVED** [review]

**Pull request closed without merging**